File: CPP/lee599.cpp
//
// Created by yangf on 2022/3/14.
//
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        unordered_map<string, int> records;
        for (int i = 0; i < list1.size(); i++) {
            records[list1[i]] = i;
        }

        vector<string> rslt;
        int result = list1.size() + list2.size();
        for (int i = 0; i < list2.size(); i++) {
            if (records.count(list2[i]) != 0) {
                int curVal = i + records[list2[i]];
                if (curVal > result) {
                    continue;
                }
                if (curVal < result) {
                    rslt.clear();
                    result = curVal;
                }
                rslt.emplace_back(list2[i]);
            }
        }
        return rslt;
    }
};

File: JAVA/Lee599.java
package JAVA;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Lee599 {
    public String[] findRestaurant(String[] list1, String[] list2) {
        Map<String, Integer> records = new HashMap<>();

        for(int i = 0; i < list1.length; i++) {
            records.put(list1[i], i);
        }

        List<String> resultList = new ArrayList<>();
        int minSum = list1.length + list2.length;
        for (int i = 0; i < list2.length; i++) {
            if (!records.containsKey(list2[i])) {
                continue;
            }
            int curSum = i + records.get(list2[i]);
            if (curSum > minSum) {
                continue;
            }
            if (curSum < minSum) {
                resultList.clear();
                minSum = curSum;
            }
            resultList.add(list2[i]);
        }
        return resultList.toArray(new String[0]);
    }
}
